evap/results/tools.py

```diff
@@ -171,42 +170,109 @@ def _calculate_results_impl(course):
 
-    final_avg = mix(final_grade_avg, final_likert_avg, settings.GRADE_PERCENTAGE)
-    final_dev = mix(final_grade_dev, final_likert_dev, settings.GRADE_PERCENTAGE)
+def calculate_average_grade_and_distribution(course):
+    # will contain a list of results for each contributor and one for the course (where contributor is None)
+    grouped_results = defaultdict(list)
+    for __, contributor, __, results, __ in calculate_results(course):
+        grouped_results[contributor].extend(results)
+
+    # early return if no rating results exist for this course
+    if not any(result.average for result in grouped_results[contributor] for contributor in grouped_results.values() if result.question.is_rating_question):
+        return None, None
+
+    contributor_averages = []
+    contributor_distributions = []
+
+    for contributor in grouped_results:
+        grade_question_results = [result for result in grouped_results[contributor] if result.question.is_grade_question]
+        grade_questions_average = avg([result.average for result in grade_question_results])
+        grade_questions_distribution = avg_distribution([result.counts for result in grade_question_results])
+        non_grade_question_results = [result for result in grouped_results[contributor] if result.question.is_non_grade_rating_question]
+        non_grade_questions_average = avg([result.average for result in non_grade_question_results])
+        non_grade_questions_distribution = avg_distribution([result.counts for result in non_grade_question_results])
+        if contributor:
+            # collect the grade question average, non-grade question average and the distributions for each contributor
+            contributor_averages.append(mix(grade_questions_average, non_grade_questions_average, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+            contributor_distributions.append(weighted_avg_distribution(
+                [grade_questions_distribution, non_grade_questions_distribution],
+                [settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE, 1 - settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE]
+            ))
+        else:
+            # calculate the grade question average, non-grade question average and the distributions for the course
+            course_grade_questions_average = grade_questions_average
+            course_grade_questions_distribution = grade_questions_distribution
+            course_non_grade_questions_average = non_grade_questions_average
+            course_non_grade_questions_distribution = non_grade_questions_distribution
+
+    if course.is_single_result:
+        course_grade_questions_average = None
+        course_non_grade_questions_average = None
+
+    # calculate the average grade and distribution over all contributors
+    contributors_average = avg(contributor_averages)
+    contributors_distribution = avg_distribution(contributor_distributions)
```

Comment:
what about calling these `contributors_average_grade` and `contributors_average_distribution` and removing the comment?